gcore/gdal_minmax_element.hpp

```diff
@@ -0,0 +1,971 @@
+            return extremum_element<T, IS_MAX>(
+                static_cast<const T *>(buffer), nElts, bHasNoData,
+                bHasNoData ? static_cast<T>(dfNoDataValue) : 0);
+        }
+        case GDT_Float64:
+        {
+            using T = double;
+            bHasNoData = bHasNoData && GDALIsValueExactAs<T>(dfNoDataValue);
+            return extremum_element<T, IS_MAX>(
+                static_cast<const T *>(buffer), nElts, bHasNoData,
+                bHasNoData ? static_cast<T>(dfNoDataValue) : 0);
+        }
+        default:
+            break;
+    }
+    CPLError(CE_Failure, CPLE_NotSupported,
+             "%s not supported for this data type.", __FUNCTION__);
+    return 0;
+}
+
+}  // namespace detail
+
+/************************************************************************/
+/*                            max_element()                             */
+/************************************************************************/
+
+/** Return the index of the element where the maximum value is hit.
+ *
+ * If it is hit in several locations, it is not specified which one will be
+ * returned.
+ *
+ * @param buffer Vector of nElts elements of type eDT.
+ * @param nElts Number of elements in buffer.
+ * @param eDT Data type of the elements of buffer.
+ * @param bHasNoData Whether dfNoDataValue is valid.
+ * @param dfNoDataValue Nodata value, only taken into account if bHasNoData == true
+ *
+ * @since GDAL 3.11
+ */
+inline size_t max_element(const void *buffer, size_t nElts, GDALDataType eDT,
+                          bool bHasNoData, double dfNoDataValue)
+{
+    return detail::extremum_element<true>(buffer, nElts, eDT, bHasNoData,
+                                          dfNoDataValue);
+}
+
+/************************************************************************/
+/*                            min_element()                             */
+/************************************************************************/
+
+/** Return the index of the element where the minimum value is hit.
+ *
+ * If it is hit in several locations, it is not specified which one will be
+ * returned.
+ *
+ * @param buffer Vector of nElts elements of type eDT.
+ * @param nElts Number of elements in buffer.
+ * @param eDT Data type of the elements of buffer.
+ * @param bHasNoData Whether dfNoDataValue is valid.
+ * @param dfNoDataValue Nodata value, only taken into account if bHasNoData == true
+ *
+ * @since GDAL 3.11
+ */
+inline size_t min_element(const void *buffer, size_t nElts, GDALDataType eDT,
+                          bool bHasNoData, double dfNoDataValue)
+{
+    return detail::extremum_element<false>(buffer, nElts, eDT, bHasNoData,
+                                           dfNoDataValue);
+}
+
+namespace detail
+{
+
+#ifdef NOT_EFFICIENT
+
+/************************************************************************/
+/*                         minmax_element()                             */
```

Comment:
So there are algorithms for finding both the min and max element of a list of numbers at the same time that beat the speed of iterating the list and checking each number versus min and each numeber versus mac and updating the min and max as you go. You can do less comparisons by pulling out two numbers at a time, comparing them to each other, and checking the min one to overall min and the max one versus overall max. I think this ends up doing 3/4's the number of overall comparisons than the naive approach. I have lost the reference for this but I coded it in my java library. See here; https://github.com/bdezonia/zorbage/blob/master/src/main/java/nom/bdezonia/zorbage/algorithm/MinMaxElement.java

Reply:
Thanks for the suggestion but saving 1/4 of operations isn't going to improve performance in a significant way.

Current timings are for a 10 million Byte vector:
- using std::minmax_element(): 15,062,444
- using NOT_EFFICIENT code path: 9,625,260
- using SSE2 optimized gdal::min_element() + gdal::max_element() (enabled currently): 1,341,901

So your suggestion would perhaps reduce NOT_EFFICIENT code path timing, but it would be still far away from the  twice repeated SSE2 optimized code paths. Actually I've just tested it and it results in... 50% slower code than the naive version... Presumably to degraded pipelining